### PR TITLE
Support JSON output from certigo

### PR DESCRIPTION
**Pull request by worldwise001**
cc: @christodenny

### CODE REVIEW

File: display.go [hunk 1]
```diff
@@ -75,7 +78,103 @@ type certWithName struct {
...
+	Organization       []string `json:"organization"`
+	OrganizationalUnit []string `json:"organizational_unit"`
+	Country            []string `json:"country"`
+	Locality           []string `json:"locality"`
+	KeyId              string   `json:"key_id,omitempty"`
+}
+
+type basicConstraints struct {
+	IsCA       bool `json:"is_ca"`
+	MaxPathLen int  `json:"pathlen"`
```
Comment (csstaub) [inline display.go]: What about if they're critical or not?

File: display.go [hunk 2]
```diff
@@ -75,7 +78,103 @@ type certWithName struct {
...
+}
+
+type nameConstraints struct {
+	Critical            bool     `json:"critical"`
+	PermittedDNSDomains []string `json:"permitted_dns_domains"`
```
Comment (csstaub) [inline display.go]: What about critical true/false?

Comment (csstaub) [inline display.go]: Also `permitted_dns_domains` might be misleading. What about negative constraints?

File: display.go [hunk 3]
```diff
@@ -75,7 +78,103 @@ type certWithName struct {
...
+}
+
+type certBlob struct {
+	Alias              string           `json:"alias,omitempty"`
+	SerialNumber       *big.Int         `json:"serial"`
+	NotBefore          int64            `json:"not_before"`
+	NotAfter           int64            `json:"not_after"`
```
Comment (csstaub) [inline display.go]: Use `time.Time` maybe?

File: display.go [hunk 4]
```diff
@@ -75,7 +78,103 @@ type certWithName struct {
 	cert *x509.Certificate
 }
 
-func displayCertFromX509(block *pem.Block) {
+type dn struct {
```
Comment (csstaub) [inline display.go]: More descriptive name here? `PKIXName`, `X500Name`, ...?

File: display.go (hunk 3 above)
Comment (worldwise001) [inline display.go]: Will that parse properly in json? How do we want the time to be output in json?

File: display.go [hunk 5]
```diff
@@ -75,7 +78,103 @@ type certWithName struct {
...
+	Issuer             dn               `json:"issuer"`
+	BasicConstraints   basicConstraints `json:"basic_constraints"`
+	NameConstraints    nameConstraints  `json:"name_constraints"`
+	KeyUsage           []string         `json:"key_usage"`
+	ExtKeyUsage        []string         `json:"extended_key_usage"`
+	AltDNSNames        []string         `json:"alternate_dns_names,omitempty"`
+	AltIPAddresses     []string         `json:"alternate_ip_addresses,omitempty"`
+	EmailAddresses     []string         `json:"email_addresses,omitempty"`
+	Warnings           []string         `json:"warnings,omitempty"`
+	original           *x509.Certificate
```
Comment (csstaub) [inline display.go]: Use embedding maybe? https://golang.org/doc/effective_go.html#embedding

File: display.go (hunk 3 above)
Comment (csstaub) [inline display.go]: ISO 8601

**Comment by christodenny** [comment]
? @worldwise001

File: cert_encoder.go [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+}
+
+type nameConstraints struct {
+	Critical            bool     `json:"critical"`
+	PermittedDNSDomains []string `json:"permitted_dns_domains,omitempty"`
+}
+
+type simpleCertificate struct {
+	Alias              string              `json:"alias,omitempty"`
+	SerialNumber       *big.Int            `json:"serial"`
```
Comment (csstaub) [inline cert_encoder.go]: The serial doesn't print correctly, gives me something like this:
``` 
  "serial": 2.6068085574204307e+38,
```

File: cert_encoder.go [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+	x509.DSAWithSHA256:   "DSA-SHA256",
+	x509.ECDSAWithSHA1:   "ECDSA-SHA1",
+	x509.ECDSAWithSHA256: "ECDSA-SHA256",
+	x509.ECDSAWithSHA384: "ECDSA-SHA384",
+	x509.ECDSAWithSHA512: "ECDSA-SHA512",
+}
+
+type basicConstraints struct {
+	IsCA       bool `json:"is_ca"`
+	MaxPathLen int  `json:"pathlen"`
```
Comment (csstaub) [inline cert_encoder.go]: We should elide pathlen if it's not set. Right now I'm getting:
```
  "pathlen": -1
```

File: cert_encoder.go [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+	Alias              string              `json:"alias,omitempty"`
+	SerialNumber       *big.Int            `json:"serial"`
+	NotBefore          time.Time           `json:"not_before"`
+	NotAfter           time.Time           `json:"not_after"`
+	SignatureAlgorithm simpleSigAlg        `json:"signature_algorithm"`
+	IsSelfSigned       bool                `json:"is_self_signed"`
+	Subject            simplePkixName      `json:"subject"`
+	Issuer             simplePkixName      `json:"issuer"`
+	BasicConstraints   basicConstraints    `json:"basic_constraints"`
+	NameConstraints    nameConstraints     `json:"name_constraints"`
```
Comment (csstaub) [inline cert_encoder.go]: Should be elided if cert doesn't have name constraints. Right now seeing:
```
  "name_constraints": {
    "critical": false,
    "permitted_dns_domains": null
  },
```

File: cert_encoder.go [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+type simpleCertificate struct {
+	Alias              string              `json:"alias,omitempty"`
+	SerialNumber       *big.Int            `json:"serial"`
+	NotBefore          time.Time           `json:"not_before"`
+	NotAfter           time.Time           `json:"not_after"`
+	SignatureAlgorithm simpleSigAlg        `json:"signature_algorithm"`
+	IsSelfSigned       bool                `json:"is_self_signed"`
+	Subject            simplePkixName      `json:"subject"`
+	Issuer             simplePkixName      `json:"issuer"`
+	BasicConstraints   basicConstraints    `json:"basic_constraints"`
```
Comment (csstaub) [inline cert_encoder.go]: Needs to be elided if empty.
```
  "basic_constraints": {
    "is_ca": false,
    "pathlen": -1
  },
```

File: cert_encoder.go [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+	out := map[string]interface{}{}
+
+	if p.Name.CommonName != "" {
+		out["common_name"] = p.Name.CommonName
+	}
+	if len(p.Name.Organization) > 0 {
+		out["organization"] = p.Name.Organization
+	}
+	if len(p.Name.OrganizationalUnit) > 0 {
+		out["organization_unit"] = p.Name.OrganizationalUnit
```
Comment (csstaub) [inline cert_encoder.go]: `s/organization_unit/organizational_unit`

File: cert_encoder.go [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+	IsCA       bool `json:"is_ca"`
+	MaxPathLen int  `json:"pathlen"`
+}
+
+type nameConstraints struct {
+	Critical            bool     `json:"critical"`
+	PermittedDNSDomains []string `json:"permitted_dns_domains,omitempty"`
+}
+
+type simpleCertificate struct {
```
Comment (csstaub) [inline cert_encoder.go]: Add fingerprint?

File: verify.go [hunk 12]
```diff
@@ -21,9 +21,23 @@ import (
...
+	SignatureAlgorithm simpleSigAlg `json:"signature_algorithm"`
+}
+
+type simpleVerifyChain struct {
+	Certs []simpleVerifyCert `json:"chain"`
+}
+
+type simpleVerification struct {
+	Error  string              `json:"error,omitempty"`
+	Chains []simpleVerifyChain `json:"chains"`
```
Comment (csstaub) [inline verify.go]: Right now, this prints as: 
```
{
  "chains": [
    {
      "chain": [
          // stuff
      ],
      "chain": [
          // stuff
      ],
```
which seems odd.

File: verify.go [hunk 13]
```diff
@@ -21,9 +21,23 @@ import (
 	"fmt"
 	"io/ioutil"
 	"os"
-	"strings"
 )
 
+type simpleVerifyCert struct {
+	Name               string       `json:"name"`
```
Comment (csstaub) [inline verify.go]: We could add serial here. Not all certs have common name.

**Comment by csstaub** [comment]
<details>
<summary>Here's some output I got for squareup.com:443</summary>
``` 
{
  "certificates": [
    {
      "serial": 2.6068085574204307e+38,
      "not_before": "2016-07-15T20:15:52Z",
      "not_after": "2017-07-31T20:45:50Z",
      "signature_algorithm": "SHA256-RSA",
      "is_self_signed": false,
      "subject": {
        "common_name": "www.squareup.com",
        "country": [
          "US"
        ],
        "key_id": "D4:17:14:6F:0B:C5:20:A1:D6:FE:21:7E:DC:9E:F8:57:9C:ED:AE:6A",
        "locality": [
          "San Francisco"
        ],
        "organization": [
          "Square, Inc."
        ]
      },
      "issuer": {
        "common_name": "Entrust Certification Authority - L1M",
        "country": [
          "US"
        ],
        "key_id": "C3:F7:D0:B5:2A:30:AD:AF:0D:91:21:70:39:54:DD:BC:89:70:C7:3A",
        "organization": [
          "Entrust, Inc."
        ],
        "organization_unit": [
          "See www.entrust.net/legal-terms",
          "(c) 2014 Entrust, Inc. - for authorized use only"
        ]
      },
      "basic_constraints": {
        "is_ca": false,
        "pathlen": -1
      },
      "name_constraints": {
        "critical": false
      },
      "key_usage": [
        "Digital Signature",
        "Key Encipherment"
      ],
      "extended_key_usage": [
        "Server Auth",
        "Client Auth"
      ],
      "alternate_dns_names": [
        "www.squareup.com",
        "squareup.com",
        "account.squareup.com",
        "mkt.com",
        "www.mkt.com",
        "market.squareup.com",
        "gosq.com",
        "www.gosq.com",
        "gosq.co",
        "www.gosq.co"
      ]
    },
    {
      "serial": 30215777750102225000000000000,
      "not_before": "2014-12-15T15:25:03Z",
      "not_after": "2030-10-15T15:55:03Z",
      "signature_algorithm": "SHA256-RSA",
      "is_self_signed": false,
      "subject": {
        "common_name": "Entrust Certification Authority - L1M",
        "country": [
          "US"
        ],
        "key_id": "C3:F7:D0:B5:2A:30:AD:AF:0D:91:21:70:39:54:DD:BC:89:70:C7:3A",
        "organization": [
          "Entrust, Inc."
        ],
        "organization_unit": [
          "See www.entrust.net/legal-terms",
          "(c) 2014 Entrust, Inc. - for authorized use only"
        ]
      },
      "issuer": {
        "common_name": "Entrust Root Certification Authority - G2",
        "country": [
          "US"
        ],
        "key_id": "6A:72:26:7A:D0:1E:EF:7D:E7:3B:69:51:D4:6C:8D:9F:90:12:66:AB",
        "organization": [
          "Entrust, Inc."
        ],
        "organization_unit": [
          "See www.entrust.net/legal-terms",
          "(c) 2009 Entrust, Inc. - for authorized use only"
        ]
      },
      "basic_constraints": {
        "is_ca": true,
        "pathlen": 0
      },
      "name_constraints": {
        "critical": false
      },
      "key_usage": [
        "Cert Sign",
        "CRL Sign"
      ],
      "extended_key_usage": [
        "Client Auth",
        "Server Auth"
      ]
    },
    {
      "serial": 1372799044,
      "not_before": "2014-09-22T17:14:57Z",
      "not_after": "2024-09-23T01:31:53Z",
      "signature_algorithm": "SHA256-RSA",
      "is_self_signed": false,
      "subject": {
        "common_name": "Entrust Root Certification Authority - G2",
        "country": [
          "US"
        ],
        "key_id": "6A:72:26:7A:D0:1E:EF:7D:E7:3B:69:51:D4:6C:8D:9F:90:12:66:AB",
        "organization": [
          "Entrust, Inc."
        ],
        "organization_unit": [
          "See www.entrust.net/legal-terms",
          "(c) 2009 Entrust, Inc. - for authorized use only"
        ]
      },
      "issuer": {
        "common_name": "Entrust Root Certification Authority",
        "country": [
          "US"
        ],
        "key_id": "68:90:E4:67:A4:A6:53:80:C7:86:66:A4:F1:F7:4B:43:FB:84:BD:6D",
        "organization": [
          "Entrust, Inc."
        ],
        "organization_unit": [
          "www.entrust.net/CPS is incorporated by reference",
          "(c) 2006 Entrust, Inc."
        ]
      },
      "basic_constraints": {
        "is_ca": true,
        "pathlen": 1
      },
      "name_constraints": {
        "critical": false
      },
      "key_usage": [
        "Cert Sign",
        "CRL Sign"
      ]
    }
  ],
  "verify_result": {
    "chains": [
      {
        "chain": [
          {
            "name": "www.squareup.com",
            "is_self_signed": false,
            "signature_algorithm": "SHA256-RSA"
          },
          {
            "name": "Entrust Certification Authority - L1M",
            "is_self_signed": false,
            "signature_algorithm": "SHA256-RSA"
          },
          {
            "name": "Entrust Root Certification Authority - G2",
            "is_self_signed": true,
            "signature_algorithm": "SHA256-RSA"
          }
        ]
      },
      {
        "chain": [
          {
            "name": "www.squareup.com",
            "is_self_signed": false,
            "signature_algorithm": "SHA256-RSA"
          },
          {
            "name": "Entrust Certification Authority - L1M",
            "is_self_signed": false,
            "signature_algorithm": "SHA256-RSA"
          },
          {
            "name": "Entrust Root Certification Authority - G2",
            "is_self_signed": false,
            "signature_algorithm": "SHA256-RSA"
          },
          {
            "name": "Entrust Root Certification Authority",
            "is_self_signed": true,
            "signature_algorithm": "SHA1-RSA"
          }
        ]
      }
    ]
  }
}
```
</details>

File: cert_encoder.go [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+	NotAfter           time.Time           `json:"not_after"`
+	SignatureAlgorithm simpleSigAlg        `json:"signature_algorithm"`
+	IsSelfSigned       bool                `json:"is_self_signed"`
+	Subject            simplePkixName      `json:"subject"`
+	Issuer             simplePkixName      `json:"issuer"`
+	BasicConstraints   basicConstraints    `json:"basic_constraints"`
+	NameConstraints    nameConstraints     `json:"name_constraints"`
+	KeyUsage           simpleKeyUsage      `json:"key_usage,omitempty"`
+	ExtKeyUsage        []simpleExtKeyUsage `json:"extended_key_usage,omitempty"`
+	AltDNSNames        []string            `json:"alternate_dns_names,omitempty"`
```
Comment (mcpherrinm) [inline cert_encoder.go]: I prefer just "dns_names" / "ip_addresses".

**Comment by csstaub** [comment]
I rolled this into #96

**Pull request closed without merging**